### ISSUE TITLE
One-time-code bad URL

### ISSUE THREAD

**Issue by palonsoro** [cloudfoundry/uaa#161]
Good evening.

My issue is quite simple: whenever I try to use `cf login --sso`, I am suggested to go to an URL like `https://uaa.MYSYSTEMDOMAIN/passcode` (using `uaa` route) to get the passcode. This URL does not work. 

However, the following URL does work: `https://login.MYSYSTEMDOMAIN/passcode` (using `login` root, inherited from the old login server). I am using v211 with no legacy login server, just the UAA.

How could this be fixed?

Thank you in advance

**Comment by fhanik** [comment]
I just tested this on:
http://uaa.run.pivotal.io/passcode

and it worked without issues. I didn't have to revert to login subdomain.

![image](https://cloud.githubusercontent.com/assets/607429/8085484/a8478554-0f90-11e5-9d39-833f60d2e83f.png)

**Comment by fhanik** [comment]
You can reopen if you provide more info (like in Chrome, record the network traffic and include it)

**Comment by palonsoro** [comment]
Ok. I forgot to mention that this happened while trying to log in via an external identity provider.
If you log in via an UAA account, it does work fine, as you have shown.
However, if you log in via an external SAML IDP, it does not.

I cannot attach a full traffic capture, because the CF instance I am working at is not mine but my company's and a traffic capture may contain sensitive data. 

However, I have followed it and saw the following: If you go to `https://uaa.MYSYSTEMDOMAIN/passcode` (or `https://uaa.MYSYSTEMDOMAIN/whatever`) and log in via the external IDP, you are redirected to `https://login.MYSYSTEMDOMAIN/login`. Then, when you want to go back to `https://uaa.MYSYSTEMDOMAIN/passcode`, you are prompted to log in again.

I have been further testing and I found out that, if you log in `https://uaa.MYSYSTEMDOMAIN/`, you are not logged in `https://login.MYSYSTEMDOMAIN` and vice versa. Furthermore, you can even log in using different users to `https://login.MYSYSTEMDOMAIN` and `https://uaa.MYSYSTEMDOMAIN/` at the same time in the same browser. So it looks like sessions are separated for both routes.

The problem is that, when logging in with an external IDP, you are always redirected to (and thus you log in to) the login endpoint, although you want to go to the uaa endpoint.

**Comment by palonsoro** [comment]
I am using cf community relase v211

**Comment by fhanik** [comment]
Saml is a two way contract. that means if your saml Idp is configured to go to login.<domain> then you should always use that. 

if you go to UAA.<domain>/passcode this gets stored in the session. but when your SAML redirects to login.<> the session cookie isn't valid (cause it was uaa.<>)

so in your case, use one domain and stick to it.

**Comment by palonsoro** [comment]
I am pretty sure that I did stick to uaa domain, but I cannot check it until tomorrow. 

However, I had a look at SAML metadata endpoint  provided at http://uaa.run.pivotal.io/saml/metadata and I get the following:

```xml
<?xml version="1.0" encoding="UTF-8"?><md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" ID="login.run.pivotal.io" entityID="login.run.pivotal.io"><ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/><ds:SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/><ds:Reference URI="#login.run.pivotal.io"><ds:Transforms><ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/><ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/></ds:Transforms><ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/><ds:DigestValue>Fli9xAAQoykNhl/GYLgV+M3+Aak=</ds:DigestValue></ds:Reference></ds:SignedInfo><ds:SignatureValue>3VrhmMzZecmNdLqTq07j+ES6InJQFdfftyi+qQmsVBxYvGwFMF/FON5RNDvbxlfBIOipxhZ/VoAcgv0YLVfLTfLFd0SPcYw3Knma0rCmNIoTJRZW+x9jxDKjwDcxQnOBPGOF5NQ53Mg1fIF5yeI+arWfzLIjeiRpDEhlrTy62cE=</ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509Certificate>MIIEJTCCA46gAwIBAgIJANIqfxWTfhpkMA0GCSqGSIb3DQEBBQUAMIG+MQswCQYDVQQGEwJVUzET
MBEGA1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEdMBsGA1UEChMUUGl2
b3RhbCBTb2Z0d2FyZSBJbmMxJDAiBgNVBAsTG0Nsb3VkIEZvdW5kcnkgSWRlbnRpdHkgVGVhbTEc
MBoGA1UEAxMTaWRlbnRpdHkuY2YtYXBwLmNvbTEfMB0GCSqGSIb3DQEJARYQbWFyaXNzYUB0ZXN0
Lm9yZzAeFw0xNTA1MTQxNzE5MTBaFw0yNTA1MTExNzE5MTBaMIG+MQswCQYDVQQGEwJVUzETMBEG
A1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEdMBsGA1UEChMUUGl2b3Rh
bCBTb2Z0d2FyZSBJbmMxJDAiBgNVBAsTG0Nsb3VkIEZvdW5kcnkgSWRlbnRpdHkgVGVhbTEcMBoG
A1UEAxMTaWRlbnRpdHkuY2YtYXBwLmNvbTEfMB0GCSqGSIb3DQEJARYQbWFyaXNzYUB0ZXN0Lm9y
ZzCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA30y2nX+kICXktl1yJhBzLGvtTuzJiLeOMWi+
+zdivifyRqX1dwJ5MgdOsBWdNrASwe4ZKONiyLFRDsk7lAYq3f975chxSsrRu1BLetBZfPEmwBH7
FCTdYtWklJbpz0vzQs/gSsMChT/UrN6zSJhPVHNizLxstedyxxVVts644U8CAwEAAaOCAScwggEj
MB0GA1UdDgQWBBSvWY/TyHysYGxKvII95wD/CzE1AzCB8wYDVR0jBIHrMIHogBSvWY/TyHysYGxK
vII95wD/CzE1A6GBxKSBwTCBvjELMAkGA1UEBhMCVVMxEzARBgNVBAgTCkNhbGlmb3JuaWExFjAU
BgNVBAcTDVNhbiBGcmFuY2lzY28xHTAbBgNVBAoTFFBpdm90YWwgU29mdHdhcmUgSW5jMSQwIgYD
VQQLExtDbG91ZCBGb3VuZHJ5IElkZW50aXR5IFRlYW0xHDAaBgNVBAMTE2lkZW50aXR5LmNmLWFw
cC5jb20xHzAdBgkqhkiG9w0BCQEWEG1hcmlzc2FAdGVzdC5vcmeCCQDSKn8Vk34aZDAMBgNVHRME
BTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAL5j1JCN5EoXMOOBSBUL8KeVZFQD3NfyYkYKBatFEKdB
FlAKLBdG+5KzE7sTYesn7EzBISHXFz3DhdK2tg+IF1DeSFVmFl2niVxQ1sYjo4kCugHBsWo+MpFH
9VBLFzsMlP3eIDuVKe8aPXFKYCGhctZEJdQTKljalshe50nayKrT</ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><md:SPSSODescriptor AuthnRequestsSigned="true" WantAssertionsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol"><md:KeyDescriptor use="signing"><ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#"><ds:X509Data><ds:X509Certificate>MIIEJTCCA46gAwIBAgIJANIqfxWTfhpkMA0GCSqGSIb3DQEBBQUAMIG+MQswCQYDVQQGEwJVUzET
MBEGA1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEdMBsGA1UEChMUUGl2
b3RhbCBTb2Z0d2FyZSBJbmMxJDAiBgNVBAsTG0Nsb3VkIEZvdW5kcnkgSWRlbnRpdHkgVGVhbTEc
MBoGA1UEAxMTaWRlbnRpdHkuY2YtYXBwLmNvbTEfMB0GCSqGSIb3DQEJARYQbWFyaXNzYUB0ZXN0
Lm9yZzAeFw0xNTA1MTQxNzE5MTBaFw0yNTA1MTExNzE5MTBaMIG+MQswCQYDVQQGEwJVUzETMBEG
A1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEdMBsGA1UEChMUUGl2b3Rh
bCBTb2Z0d2FyZSBJbmMxJDAiBgNVBAsTG0Nsb3VkIEZvdW5kcnkgSWRlbnRpdHkgVGVhbTEcMBoG
A1UEAxMTaWRlbnRpdHkuY2YtYXBwLmNvbTEfMB0GCSqGSIb3DQEJARYQbWFyaXNzYUB0ZXN0Lm9y
ZzCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA30y2nX+kICXktl1yJhBzLGvtTuzJiLeOMWi+
+zdivifyRqX1dwJ5MgdOsBWdNrASwe4ZKONiyLFRDsk7lAYq3f975chxSsrRu1BLetBZfPEmwBH7
FCTdYtWklJbpz0vzQs/gSsMChT/UrN6zSJhPVHNizLxstedyxxVVts644U8CAwEAAaOCAScwggEj
MB0GA1UdDgQWBBSvWY/TyHysYGxKvII95wD/CzE1AzCB8wYDVR0jBIHrMIHogBSvWY/TyHysYGxK
vII95wD/CzE1A6GBxKSBwTCBvjELMAkGA1UEBhMCVVMxEzARBgNVBAgTCkNhbGlmb3JuaWExFjAU
BgNVBAcTDVNhbiBGcmFuY2lzY28xHTAbBgNVBAoTFFBpdm90YWwgU29mdHdhcmUgSW5jMSQwIgYD
VQQLExtDbG91ZCBGb3VuZHJ5IElkZW50aXR5IFRlYW0xHDAaBgNVBAMTE2lkZW50aXR5LmNmLWFw
cC5jb20xHzAdBgkqhkiG9w0BCQEWEG1hcmlzc2FAdGVzdC5vcmeCCQDSKn8Vk34aZDAMBgNVHRME
BTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAL5j1JCN5EoXMOOBSBUL8KeVZFQD3NfyYkYKBatFEKdB
FlAKLBdG+5KzE7sTYesn7EzBISHXFz3DhdK2tg+IF1DeSFVmFl2niVxQ1sYjo4kCugHBsWo+MpFH
9VBLFzsMlP3eIDuVKe8aPXFKYCGhctZEJdQTKljalshe50nayKrT</ds:X509Certificate></ds:X509Data></ds:KeyInfo></md:KeyDescriptor><md:KeyDescriptor use="encryption"><ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#"><ds:X509Data><ds:X509Certificate>MIIEJTCCA46gAwIBAgIJANIqfxWTfhpkMA0GCSqGSIb3DQEBBQUAMIG+MQswCQYDVQQGEwJVUzET
MBEGA1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEdMBsGA1UEChMUUGl2
b3RhbCBTb2Z0d2FyZSBJbmMxJDAiBgNVBAsTG0Nsb3VkIEZvdW5kcnkgSWRlbnRpdHkgVGVhbTEc
MBoGA1UEAxMTaWRlbnRpdHkuY2YtYXBwLmNvbTEfMB0GCSqGSIb3DQEJARYQbWFyaXNzYUB0ZXN0
Lm9yZzAeFw0xNTA1MTQxNzE5MTBaFw0yNTA1MTExNzE5MTBaMIG+MQswCQYDVQQGEwJVUzETMBEG
A1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEdMBsGA1UEChMUUGl2b3Rh
bCBTb2Z0d2FyZSBJbmMxJDAiBgNVBAsTG0Nsb3VkIEZvdW5kcnkgSWRlbnRpdHkgVGVhbTEcMBoG
A1UEAxMTaWRlbnRpdHkuY2YtYXBwLmNvbTEfMB0GCSqGSIb3DQEJARYQbWFyaXNzYUB0ZXN0Lm9y
ZzCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA30y2nX+kICXktl1yJhBzLGvtTuzJiLeOMWi+
+zdivifyRqX1dwJ5MgdOsBWdNrASwe4ZKONiyLFRDsk7lAYq3f975chxSsrRu1BLetBZfPEmwBH7
FCTdYtWklJbpz0vzQs/gSsMChT/UrN6zSJhPVHNizLxstedyxxVVts644U8CAwEAAaOCAScwggEj
MB0GA1UdDgQWBBSvWY/TyHysYGxKvII95wD/CzE1AzCB8wYDVR0jBIHrMIHogBSvWY/TyHysYGxK
vII95wD/CzE1A6GBxKSBwTCBvjELMAkGA1UEBhMCVVMxEzARBgNVBAgTCkNhbGlmb3JuaWExFjAU
BgNVBAcTDVNhbiBGcmFuY2lzY28xHTAbBgNVBAoTFFBpdm90YWwgU29mdHdhcmUgSW5jMSQwIgYD
VQQLExtDbG91ZCBGb3VuZHJ5IElkZW50aXR5IFRlYW0xHDAaBgNVBAMTE2lkZW50aXR5LmNmLWFw
cC5jb20xHzAdBgkqhkiG9w0BCQEWEG1hcmlzc2FAdGVzdC5vcmeCCQDSKn8Vk34aZDAMBgNVHRME
BTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAL5j1JCN5EoXMOOBSBUL8KeVZFQD3NfyYkYKBatFEKdB
FlAKLBdG+5KzE7sTYesn7EzBISHXFz3DhdK2tg+IF1DeSFVmFl2niVxQ1sYjo4kCugHBsWo+MpFH
9VBLFzsMlP3eIDuVKe8aPXFKYCGhctZEJdQTKljalshe50nayKrT</ds:X509Certificate></ds:X509Data></ds:KeyInfo></md:KeyDescriptor><md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://login.run.pivotal.io/saml/SingleLogout/alias/login.run.pivotal.io"/><md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://login.run.pivotal.io/saml/SingleLogout/alias/login.run.pivotal.io"/><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:transient</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:persistent</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:X509SubjectName</md:NameIDFormat><md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://login.run.pivotal.io/saml/SSO/alias/login.run.pivotal.io" index="0" isDefault="true"/><md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact" Location="https://login.run.pivotal.io/saml/SSO/alias/login.run.pivotal.io" index="1"/></md:SPSSODescriptor></md:EntityDescriptor>
```

As you can see, it redirects to login.run.pivotal.io so, if the same happens at community CF, you cannot stick to uaa endpoint at all. However, you are always suggested to use the uaa endpoint by the CLI.

**Comment by fhanik** [comment]
HTTP-POST" Location="https://login.run.pivotal.io/saml/SingleLogout/alias/login.run.pivotal.io"

this is the default setup, with no changes. but run.pivotal.io does not have a corresponding IDP configured at all. Meaning, there is no SAML authentication there. So run.pivotal.io has not setup SAML, it just has accepted the defaults

In your case, you would configure your UAA job (through CF manifest) to use the domain you want to use.

**Comment by palonsoro** [comment]
I know that there is no SAML authentication in run.pivotal.io. I only wanted to demonstrate that SAML service provider metadata were generated pointing to the login endpoint.

I did set `uaa.url` to the uaa endpoint. I have set `login.url` to the UAA endpoint, but it did not fix this. 

Having a deeper look, I see that at [login.yml template](https://github.com/cloudfoundry/cf-release/blob/master/jobs/uaa/templates/login.yml.erb), I see that the key `login.entityBaseURL` is filled with `<%= "#{protocol}://login.#{properties.domain}" %>`, which is the login endpoint. 

Looking at the UAA code, I see at file [uaa/uaa/src/main/webapp/WEB-INF/spring/saml-providers.xml](https://github.com/cloudfoundry/uaa/blob/091c5e5961dd33c8c7ca5a15f4020e47d266a1c3/uaa/src/main/webapp/WEB-INF/spring/saml-providers.xml), I see the following snippet:

```xml
<!-- Filter automatically generates default SP metadata -->
        <bean id="metadataGeneratorFilter" class="org.springframework.security.saml.metadata.MetadataGeneratorFilter">
            <constructor-arg>
                <bean class="org.cloudfoundry.identity.uaa.login.saml.ZoneAwareMetadataGenerator">
                    <property name="extendedMetadata" ref="extendedMetaData"/>
                    <property name="requestSigned" value="${login.saml.signRequest:true}" />
                    <property name="wantAssertionSigned" value="${login.saml.wantAssertionSigned:false}" />
                    <property name="entityBaseURL" value="${login.entityBaseURL:http://localhost:8080/uaa}" />
                    <property name="entityId" ref="samlEntityID" />
                </bean>
            </constructor-arg>
        </bean>
```

If you have a look at [Spring security documentation](http://docs.spring.io/spring-security-saml/docs/current/reference/html/configuration-metadata.html), you can see that the property `entityBaseURL` is the base for all the SAML-related endpoints (you can see it at table 7.1, which describes properties for `org.springframework.security.saml.metadata.MetadataGenerator` bean, the superclass of the `org.cloudfoundry.identity.uaa.login.saml.ZoneAwareMetadataGenerator` used at the UAA).

So, the base URL for SAML-related endpoints comes from the `login.entityBaseURL` key at UAA `login.yml` configuration file, which is filled with the login endpoint at the `login.yml.erb` template of the CF release. Thus, all SAML endpoints will be always generated pointing to the login endpoint.

However, the message is always generated pointing to the uaa endpoint.

**Comment by fhanik** [comment]
Yes, you are correct, there are a few places where the endpoint is hardcoded to login.<domain> in cf-release project (not in the UAA itself). there is a reason for that, it was always the goto URL for SAML.

Now historically, the 'cf' cli has always used login.<domain> since that is where autologin and passcode endpoints, thus that is where it should be redirecting.

The fact that cf login -sso returns uaa.<domain>, is news to me. Since in the past it always returned login.<domain> since that is where SAML integration has ALWAYS lived. 

Note, the SAML endpoint has never changed in the history of the project. 

I guess the fix here is to remove the hard coded login.<domain> from the ERB files and allow that to be configurable through login.url

**Comment by fhanik** [comment]
I've restated the problem in this story in the public backlog
https://www.pivotaltracker.com/n/projects/997278/stories/96567690

**Comment by palonsoro** [comment]
First, I can confirm that I have been using the uaa endpoint for SAML, although it does not matter because of what I told in my previous comment.

Then, I have found a solution doing the opposite: Instead of trying to fix SAML redirection endpoints to point to the uaa endpoint, I managed to change the passcode URL suggested by the `cf login --sso` so that it points to the login endpoint.

I have been able to do it by setting `login.uaa_base` property to point to the login endpoint (that is quite ironic).

This solved this problem at the moment, but I am not sure whether other problems could appear from this change or not (I hope that it does not).

**Comment by fhanik** [comment]
That's a good workaround, and a fully valid one. I don't see any issues that could arise from that fix.

**Comment by vishweshh** [comment]
Filip and palonsoro,
You mentioned the OTP prompt shouldn't appear when SAML provider hasn't been configured.

But I'm getting the One Time Passcode prompt even though I haven't configured a SAML provider.

Here is what I did, I just got the UAA bundle from github, deployed it on one of the local server using ./gradlew run, installed cf-uaac, created a user from admin. When I try to get the token with this command uaac --t token get <user> <password> The system throws me a prompt One Time Code (Get one at http://localhost:8080/uaa/passcode):

This is not reachable from browser, I always get "Something went amiss." page

Can one of you let me know, how to suppress this 'one time passcode' prompt when SAML provider isn't configured? Appreciate your assistance, thanks.
